roxie/ccd/ccdfile.cpp

```diff
@@ -587,24 +600,200 @@ static void appendRemoteLocations(IPartDescriptor *pdesc, StringArray &locations
+    }
+    inline bool operator < ( const CacheInfoEntry &l) const { return u < l.u; }
+    inline bool operator <= ( const CacheInfoEntry &l) const { return u <= l.u; }
+    inline void operator++ () { b.page++; }
+
+    static constexpr unsigned pageBits = 13;  // 8k 'pages'
+};
+
+class CacheReportingBuffer : public CInterfaceOf<ICacheInfoRecorder>
+{
+    // A circular buffer recording recent file activity. Note that noteRead() and clear() may be called from multiple threads
+    // (other functions are assumed single-threaded) and that locking is kept to a minimum, even if it means information may be slightly inaccurate.
+    CacheInfoEntry *recentReads = nullptr;
+    std::atomic<unsigned> recentReadHead = {0};
+    unsigned recentReadSize;
+
+public:
+
+    CacheReportingBuffer(offset_t trackSize)
+    {
+        recentReadSize = trackSize >> CacheInfoEntry::pageBits;
+        if (traceLevel)
+            DBGLOG("Creating CacheReportingBuffer with %d elements", recentReadSize);
+        assertex(recentReadSize);
+        recentReads = new CacheInfoEntry[recentReadSize];
+        recentReadHead = 0;
+    }
+    CacheReportingBuffer(const CacheReportingBuffer &from)
+    {
+        // NOTE - from may be updated concurrently - we do not want to lock it
+        // There are therefore races in here, but they do not matter (may result in very recent data being regarded as very old or vice versa).
+        recentReadSize = from.recentReadSize;
+        recentReadHead = from.recentReadHead.load(std::memory_order_relaxed);
+        recentReads = new CacheInfoEntry[recentReadSize];
+        memcpy(recentReads, from.recentReads, recentReadSize * sizeof(CacheInfoEntry));
+    }
+
+    ~CacheReportingBuffer()
+    {
+        delete [] recentReads;
+    }
+
+    void clear()
+    {
+        recentReadHead = 0;
+    }
+
+    void noteRead(unsigned fileIdx, offset_t pos, unsigned len, bool diskCache)
+    {
+        if (recentReads && len)
+        {
+            CacheInfoEntry start(fileIdx, pos, diskCache);
+            CacheInfoEntry end(fileIdx, pos+len-1, diskCache);
+            for(;start <= end; ++start)
+            {
+                recentReads[recentReadHead++ % recentReadSize] = start;
+            }
+        }
+    }
+
+    void sortAndDedup()
+    {
+        // NOTE: single-threaded
+        unsigned sortSize;
+        if (recentReadHead > recentReadSize)
+            sortSize = recentReadSize;
+        else
+            sortSize = recentReadHead;
+        std::sort(recentReads, recentReads + sortSize);
+        CacheInfoEntry lastPos(-1,-1,false);
+        unsigned dest = 0;
+        for (unsigned idx = 0; idx < sortSize; idx++)
+        {
+            CacheInfoEntry pos = recentReads[idx];
+            if (pos.b.file != lastPos.b.file || pos.b.page != lastPos.b.page)   // Ignore inNodeCache bit when deduping
+            {
+                recentReads[dest++] = pos;
+                lastPos = pos;
+            }
+        }
```

Comment:
@richardkchapman should this have

else if (!pos.b.diskCache)
   recentReadspdest-1].b.diskCache = false;

?

Reply:
I think non disk cache will always be sorted first, so this is not necessary..